// @bit/mui-org.material-ui.internal.overridable-component/dist/OverridableComponent/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "OverridableComponent", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.OverridableComponent;
  }
});
Object.defineProperty(exports, "OverrideProps", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.OverrideProps;
  }
});
Object.defineProperty(exports, "DefaultComponentProps", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.DefaultComponentProps;
  }
});
Object.defineProperty(exports, "BaseProps", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.BaseProps;
  }
});
Object.defineProperty(exports, "CommonProps", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.CommonProps;
  }
});
Object.defineProperty(exports, "OverridableTypeMap", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.OverridableTypeMap;
  }
});
Object.defineProperty(exports, "Simplify", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.Simplify;
  }
});
Object.defineProperty(exports, "SimplifiedPropsOf", {
  enumerable: true,
  get: function get() {
    return _OverridableComponent.SimplifiedPropsOf;
  }
});

var _OverridableComponent = require("./OverridableComponent");
