// @bit/campgladiator.cgui.components.atoms.button/dist/components/atoms/Button/Button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _media = _interopRequireDefault(require("../../../utilities/media"));

var _mods = _interopRequireDefault(require("../../../utilities/mods"));

var _Icon = _interopRequireDefault(require("../Icon"));

var _Loader = _interopRequireDefault(require("../Loader"));

require("./Button.scss");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var Button = function Button(_ref) {
  var className = _ref.className,
      id = _ref.id,
      primary = _ref.primary,
      secondary = _ref.secondary,
      solid = _ref.solid,
      outline = _ref.outline,
      dark = _ref.dark,
      gray = _ref.gray,
      white = _ref.white,
      loading = _ref.loading,
      size = _ref.size,
      _ref$media = _ref.media,
      media = _ref$media === void 0 ? {} : _ref$media,
      disabled = _ref.disabled,
      children = _ref.children,
      onClick = _ref.onClick,
      icon = _ref.icon,
      props = _objectWithoutProperties(_ref, ["className", "id", "primary", "secondary", "solid", "outline", "dark", "gray", "white", "loading", "size", "media", "disabled", "children", "onClick", "icon"]);

  return _react.default.createElement("button", _extends({
    className: (0, _mods.default)('cg-button', {
      primary: primary,
      secondary: secondary,
      solid: solid,
      outline: outline,
      dark: dark,
      gray: gray,
      white: white
    }, size && "cg-button--".concat(size), _media.default.toModifiers('cg-button', media), loading && 'cg-button--loading', className),
    disabled: disabled,
    onClick: onClick,
    id: id
  }, props), icon && _react.default.createElement(_Icon.default, {
    className: "cg-button__icon",
    name: icon
  }), loading ? _react.default.createElement(_Loader.default, {
    white: solid
  }) : children);
};

var Size = _propTypes.default.oneOf(['xsmall', 'small', 'large', 'xlarge']);

Button.defaultProps = {
  className: '',
  id: '',
  icon: '',
  primary: false,
  secondary: false,
  solid: false,
  outline: false,
  dark: false,
  gray: false,
  white: false,
  loading: false,
  disabled: false
};
Button.propTypes = {
  className: _propTypes.default.string,
  id: _propTypes.default.string,
  icon: _propTypes.default.string,
  primary: _propTypes.default.bool,
  secondary: _propTypes.default.bool,
  solid: _propTypes.default.bool,
  outline: _propTypes.default.bool,
  dark: _propTypes.default.bool,
  gray: _propTypes.default.bool,
  white: _propTypes.default.bool,
  size: Size,
  loading: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  children: _propTypes.default.node,
  media: _media.default.propTypeFor(Size)
};
var _default = Button;
exports.default = _default;
